sense-android-library/src/nl/sense_os/datastorageengine/CommonSenseProxy.java

```diff
@@ -0,0 +1,597 @@
+     */
+    public static final String RESPONSE_CODE = "http response code";
+    /**
+     * Key for getting the response content from the Map object that is returned by
+     * {@link nl.sense_os.datastorageengine.CommonSenseProxy#request(String, JSONObject, String, String)}
+     */
+    public static final String RESPONSE_CONTENT = "content";
+
+    /**
+     * Default constructor for the DSECommonSenseProxy.
+     *
+     * Takes an app key that will be used throughout the proxies lifetime. Needs to know whether to talk to the live server or the staging server.
+     * This cannot be changed during the proxies lifetime. If you need to change this you simply init a new commonsense proxy.
+     *
+     * @param useLiveServer	If true, the live server will be used. If false, the staging server will be used.
+     * @param theAppKey		An application key that identifies the application to the commonsense server. Cannot be empty.
+     */
+    public CommonSenseProxy (boolean useLiveServer, String theAppKey)
+    {
+        APP_KEY = theAppKey;
+        if(useLiveServer) {
+            URL_BASE = BASE_URL_LIVE;
+            URL_AUTH = BASE_URL_AUTHENTICATION_LIVE;
+        } else {
+            URL_BASE = BASE_URL_STAGING;
+            URL_AUTH = BASE_URL_AUTHENTICATION_STAGING;
+        }
+    }
+
+
+    /**
+     * Login a user
+     *
+     * If a user is already logged in this call will also return the session ID.
+     *
+     * @param username		A user account in commonsense is uniquely identified by a username. Cannot be empty.
+     * @param password		A password in commonsense does not have any specific requirements.
+     *                      It will be MD5 hashed before sending to the server so the user does not have to provide a hashed password. Cannot be empty.
+     * @throws IOException and JSONException
+     * @return				Session ID. Will be null if the call fails.
+     */
+    public String loginUser( String username, String password) throws IOException, JSONException
+    {
+        if(username == null || username.isEmpty() || password == null || password.isEmpty())
+            throw new IOException("invalid input of username or password");
+
+        final String url = URL_AUTH + "/" + URL_LOGIN;
+        final JSONObject user = new JSONObject();
+        user.put("username", username);
+        user.put("password", password);
+
+        Map<String, String> response = request(url, user, null, HTTP_METHOD_POST);
+        // if response code is not 200 (OK), the login was incorrect
+        int result = checkResponseCode(response.get(RESPONSE_CODE), "login");
+
+        String session_id = response.get("session-id");
+
+        if (result != 0 && session_id == null) {
+            // something went horribly wrong
+            Log.w(TAG, "CommonSense login failed: no session id received!");
+            throw new IOException("login failed with no session id");
```

Comment:
@excaliburnan do we want to make a distinction between login failed due to wrong credentials and io/connection errors? Then throw different exceptions, or use parseable error codes in the return value of the exception